compiler-core/src/javascript/expression.rs

```diff
@@ -515,12 +540,45 @@ impl<'module> Generator<'module> {
             .chain(assignments.iter().map(|(assignment, _kind)| assignment));
 
         for assignment in all_assignments {
-            documents.push(self.not_in_tail_position(|gen| {
-                gen.simple_variable_assignment(&assignment.name, &assignment.value)
-            })?);
+            match assignment.value.as_ref() {
+                // An echo in a pipeline won't result in an assignment, instead it
+                // just prints the previous variable assigned in the pipeline.
+                TypedExpr::Echo {
+                    expression: None,
+                    location,
+                    ..
+                } => documents.push(self.not_in_tail_position(|gen| {
+                    let var = gen
+                        .latest_local_var
```

Comment:
Why is this state on the generator when it's specific to a pipeline? This would be the first time that function specific state is kept at the top level.